RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc

```diff
@@ -164,6 +164,400 @@ namespace lowptgsfeleid {
+                float dphi = std::fabs(ecal_pos.phi() - cluster->phi());
+                if (dphi > pi_)
+                  dphi -= 2 * pi_;
+                if (ecal_pos.phi() - cluster->phi() < 0)
+                  dphi = -dphi;
+                if (ecal_pos.eta() - cluster->eta() < 0)
+                  deta = -deta;
+
+                if (clusNum == i1) {
+                  sc_clus1_E = cluster->energy();
+                  if (gsf->pMode() > 0)
+                    sc_clus1_E_ov_p = cluster->energy() / gsf->pMode();
+                  sc_clus1_nxtal = (int)cluster->size();
+                  if (reach_ECAL > 0) {
+                    sc_clus1_deta = deta;
+                    sc_clus1_dphi = dphi;
+                  }
+                } else if (clusNum == i2) {
+                  sc_clus2_E = cluster->energy();
+                  if (gsf->pMode() > 0)
+                    sc_clus2_E_ov_p = cluster->energy() / gsf->pMode();
+                  if (reach_ECAL > 0) {
+                    sc_clus2_deta = deta;
+                    sc_clus2_dphi = dphi;
+                  }
+                }
+                clusNum++;
+              }
+            }
+          } catch (...) {
+            std::cout << "caught an exception" << std::endl;
+          }
+        }
+      }
+    }  // clusters
+
+    // Out-of-range
+    if (eid_rho < 0)
+      eid_rho = 0;
+    if (eid_rho > 100)
+      eid_rho = 100;
+    if (eid_sc_eta < -5)
+      eid_sc_eta = -5;
+    if (eid_sc_eta > 5)
+      eid_sc_eta = 5;
+    if (eid_shape_full5x5_r9 < 0)
+      eid_shape_full5x5_r9 = 0;
+    if (eid_shape_full5x5_r9 > 2)
+      eid_shape_full5x5_r9 = 2;
+    if (eid_sc_etaWidth < 0)
+      eid_sc_etaWidth = 0;
+    if (eid_sc_etaWidth > 3.14)
+      eid_sc_etaWidth = 3.14;
+    if (eid_sc_phiWidth < 0)
+      eid_sc_phiWidth = 0;
+    if (eid_sc_phiWidth > 3.14)
+      eid_sc_phiWidth = 3.14;
+    if (eid_shape_full5x5_HoverE < 0)
+      eid_shape_full5x5_HoverE = 0;
+    if (eid_shape_full5x5_HoverE > 50)
+      eid_shape_full5x5_HoverE = 50;
+    if (eid_trk_nhits < -1)
+      eid_trk_nhits = -1;
+    if (eid_trk_nhits > 50)
+      eid_trk_nhits = 50;
+    if (eid_trk_chi2red < -1)
+      eid_trk_chi2red = -1;
+    if (eid_trk_chi2red > 50)
+      eid_trk_chi2red = 50;
+    if (eid_gsf_chi2red < -1)
+      eid_gsf_chi2red = -1;
+    if (eid_gsf_chi2red > 100)
+      eid_gsf_chi2red = 100;
+    if (eid_brem_frac < 0)
+      eid_brem_frac = -1;
+    if (eid_brem_frac > 1)
+      eid_brem_frac = 1;
+    if (eid_gsf_nhits < -1)
+      eid_gsf_nhits = -1;
+    if (eid_gsf_nhits > 50)
+      eid_gsf_nhits = 50;
+    if (eid_match_SC_EoverP < 0)
+      eid_match_SC_EoverP = 0;
+    if (eid_match_SC_EoverP > 100)
+      eid_match_SC_EoverP = 100;
+    if (eid_match_eclu_EoverP < -0.001)
+      eid_match_eclu_EoverP = -0.001;
+    if (eid_match_eclu_EoverP > 0.001)
+      eid_match_eclu_EoverP = 0.001;
+    eid_match_eclu_EoverP = eid_match_eclu_EoverP * 1.E7;
```

Comment:
the value here is computed from `       eid_match_eclu_EoverP = (1. / ele->ecalEnergy()) - (1. / ele->p());`
are the ranges correct? it seems like the only way to fit in the range is to have TeV scale values.

Reply:
ehm, I guess this also includes cases where `ecalEnergy()` is very close to `p()` by construction. as these are "lowPt", e.g. \~5 GeV each would have to differ by less than \0.5%.
 rescaling this by 1E7 would mean that the MVA is supposed to make sense out of ppm difference between the two at numerical precision level.
What's the physics motivation?

Reply:
@crovelli, please can you reply?

Reply:
As you wrote, E and p are very close. A typical distribution for this variable is 
https://crovelli.web.cern.ch/crovelli/Analysis/BPark/variables_slim/eid_match_eclu_EoverP_EE.png                                                          so the range should be ok. For what concern the rescaling by 1E7, we introduced it to help the MVA training at the beginning (to have values around 1) and make the training procedure more stable from the numerical point of view. If you think it would be better not to rescale we're ok with that as well @cavalari

Reply:
Thank you for the plot.
I see that the RMS of the variable is around O(1E-7).
Please clarify if the difference is driven by physical sources (with some back of the envelope quantification, e.g. "at 5 GeV measured/published electron track resolution is 0.3% and  the ECAL resolution/calibration is within 0.2%" [I made up the % values to match the range] ), or is it numerical (something that normally should be zero but some math precision misaligned).
For the latter I think that there is no good justification to use the variable in training.

Reply:
Oh, BTW, is the plot for some J/psi electrons or a similar kinematic range? (or is it somehow for Zee)

Reply:
Hi Slava, it took a while to reply because your questions triggered some further checks... Actually, we realized that the lines https://github.com/cms-sw/cmssw/blob/644c0166533ddccc61d1e08b6d47fbe082b26bde/RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronFeatures.cc#L296-#L298 (and also the plot sent this morning) were a refusal of an old old version of the code. Very sorry for this.  The correct version of these lines is this one, which is the one currently in usage for the training: https://github.com/CMSBParking/cmssw/blob/from-CMSSW_10_2_15_2020Sept15/RecoEgamma/EgammaElectronProducers/src/LowPtGsfElectronIDExtraHeavyObjectCache.cc#L263-#L264                                                                                                 As you may see, the variable range is -1, +1 - without any rescaling. This makes sense with a back of the envelope calculation. At E~5 GeV, we can assume an ECAL resolution around 20-30% and a tracker resolution around 1%, meaning that the uncertainty on 1/E-1/p is ~0.04-0.06. This is covered by the [-1,1] range. It is also coherent with the (current, updated) version of the plot, which shows the variable for signal electrons: https://crovelli.web.cern.ch/crovelli/pippo2/eclus_gen4to6_eEcalgt2.png  This plot is done on a JPsi MC sample, we restrict to gen pT 4-6 GeV, and we ask the energy in ECAL to be >2 GeV, to remove bad reconstructed energies in ECAL.

Reply:
thank you for the back-of-the-envelope estimate and the updated plot. This looks like the truncation to 0.001 currently applied in the code is wrong, as well as the scaling by 1E7.

Reply:
Now clamp to +/-1 and removed 1E7 multiplier, https://github.com/cms-sw/cmssw/pull/31220/commits/b40aa92db1278e16f5a629e583bae36d02b18493

Reply:
IIRC, this is the only physics change. Is the slimmedLowPt selection still valid?

Reply:
Yes you are correct that this is the only physics change (along with using the B field). 

We need to retrain the BDT and retune the threshold for the selected electrons, but this will be done in a separate PR. Does this answer your question?

Reply:
> Does this answer your question?

partly.
What is the timeline for the retraining?
If the pre-selection fraction currently implemented in this PR is roughly unchanged, I'm fine for an improvement/tuning PR later.
However, if after this change we now get x10 more electrons, things are more complicated.

Reply:
This bug was only introduced within this PR (and now it is corrected).

The bug was not present for the training of the existing 2020Sept15 "placeholder" model that forms part of this PR. I did some quick tests. 

1) I compared the ID values before and after these very latest changes, using the 25202.0_TTbar wf, and found no differences for the few events I inspected. 

2) I did a quick test using `compareProducts.sh` and the 136.88811_RunJetHT2018D wf for files produced before and after these very latest changes. The increase in footprint is unchanged w.r.t. earlier studies, i.e. ~1% increase:
```
~/tools/compareProducts.sh step2_PAT_default.root step2_PAT_skimmed.root _ 1 1
Checking process _ step2_PAT_default.root and step2_PAT_skimmed.root with useUnpacked=no (if above 1 or 1%):
Compare packed values
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch
-----------------------------------------------------------------
      0.0 ->       146.6        147     NEWO   0.24     recoConversions_gsfTracksOpenConversions_gsfTracksOpenConversions_RECO.
      0.0 ->       572.7        573     NEWO   0.95     patElectrons_slimmedLowPtElectrons__PAT.
-------------------------------------------------------------
    60031 ->       60751        720             1.2     ALL BRANCHES
```

So the bug does not reveal significant changes in the BDT score, indeed they are very negligible, and certainly not an order of magnitude. 

Before we retrain the ID, we need to update the energy regression. This has started but we are awaiting the (high priority) transfer of MC samples from tape to disk. Once the transfer is complete, it should take 2-3 days to complete the regression studies and prepare an updated GlobalTag. 

Once the regression is complete, we will retrain the ID, which should take 1-2 days. The updated regression and ID are based on 10_6 AOD samples and will be performed within a 10_6 release, aimed at the UL. We'd hope to have everything ready by early next week. 

I would prefer to merge this PR before opening:
- a new PR for the energy regression
- a new PR for the retrained ID and updated threshold.
Conceivably, we could combine these PRs into a single one. 

Our priority is the UL campaign. If limited by time, we can use the regression and ID retraining done with 10_6 and integrate them in 11_2: the performance should be comparable for both release cycles.